design-docs/modes.md

```diff
@@ -0,0 +1,638 @@
+
+The enabled "Init" step in the 2025 FTC SDK/DS is not a requirement due to anticipated rule changes.  However, performing initialization of modes while the robot is still disabled is a requirement for both FTC and FRC, as it's important for user code to be able to do expensive mode-specific operations (e.g. computing autonomous paths) prior to actually starting the match.
+
+Providing the top-level teleop, autonomous, and test selection available in the 2025 FRC driver station is desirable and has few downsides for either program.  In combination with code specifying the applicable modes (also desirable), this allows for filtering down of selectable modes.  The "practice" mode feature of the 2025 FRC driver station which automatically transitions between modes based on time is a very useful tool for teams to simulate match transitions at home and should be retained, but should be renamed to reduce confusion, particularly as this mode is likely to be used for FTC matches without a FMS, "match" mode could be a good name for this feature.
+
+# Design
+
+## Overview / Key Features
+
+- DS provides a prominent enable/disable state control and a teleop/auto/test/match selector.  For teleop/auto/test, enabling the robot immediately starts execution of the selected mode; in match mode, a match sequence is followed (auto mode followed by a disabled delay, followed by teleop mode).
+
+- DS provides drop-down selector(s) for the user-defined modes; these are filtered based on the top-level mode selector (e.g. if auto is selected, a single drop-down selector of just the auto modes is provided).  For match mode or when FMS-attached, two drop downs are provided (one for auto mode selection and one for teleop mode selection).  The drop-down selector provides grouped categories as specified by the robot program.
+
+- Robot programs are structured to have a top-level Robot class (e.g. motors/sensors/subsystems); this is also where robot-wide initialization is performed (in the class constructor or static initializer block).  The Robot class provides overrideable periodic type functions for users to run code when the robot is disabled.
+
+- Modes are usually registered via annotation of classes.  These annotations may specify a name (or default to the class name), group name (for grouping of routines in the selection list), and description.  Annotations are used to specify whether the class is a teleop, auto, or test mode.
+
+- Modes may also be registered via annotation of functions (in the Robot class only) or via explicit function calls.  As C++ does not support annotations, function call registration is the only available method in that language.
+
+- For maximum flexibility, all code in the robot project has access to the enable/disable state, the overall robot teleop/auto/test mode, and the selected modes for teleop, auto, and test (even when the robot is disabled).
```

Comment:
Why is this a good thing?   It feels like this could complicate things quite a bit.

Reply:
This is important for subsystems or other facilities (e.g. logging) that might want to change behavior based on the state of the robot.  We don't want to (and can't) hide this from teams, so we should provide a good API for it.

Reply:
It might also be worth clarifying that this is read-only access, so the complexity should be minimal. The library will already need to parse this data, so unless I'm missing something, this would only involve writing convenience methods that let teams read that data.

Reply:
And in general, the library itself will use these same methods (e.g. to implement the mode transition machinery).  So it's not any additional work over what we already have to implement.